django_s3_express_cache/middleware.py

```diff
@@ -0,0 +1,254 @@
+
+
+def get_cache_key_s3_compatible(
+    request, key_prefix=None, method="GET", cache=None
+):
+    """
+    Mirrors Django's `get_cache_key` but generates a key compatible with
+    S3ExpressCacheBackend by:
+
+      • placing key_prefix at the beginning of the key
+      • replacing '/' with '.' to avoid creating nested S3 folders
+
+    If there isn't a headerlist stored, return None, indicating that the page
+    needs to be rebuilt.
+    """
+    key_prefix = key_prefix or settings.CACHE_MIDDLEWARE_KEY_PREFIX
+    cache = cache or caches[settings.CACHE_MIDDLEWARE_ALIAS]
+
+    header_key = _generate_cache_header_key_s3_compatible(key_prefix, request)
+    headerlist = cache.get(header_key)
+    if headerlist is None:
+        return None
+
+    return _generate_cache_key_s3_compatible(
+        request, method, headerlist, key_prefix
+    )
+
+
+def learn_cache_key_s3_compatible(
+    request, response, cache_timeout=None, key_prefix=None, cache=None
+):
+    """
+    Store the list of headers from the response's Vary header and generate
+    an S3-compatible cache key.
+
+    This mirrors Django's `learn_cache_key` but ensures keys are compatible
+    with S3ExpressCacheBackend:
+
+      • key_prefix appears at the start of the key (for time-based prefixing)
+      • '/' characters are replaced with '.' to prevent nested folder.
+      • all other behavior (Vary handling, sorting, i18n suffix) remains
+        the same as Django
+    """
+    key_prefix = key_prefix or settings.CACHE_MIDDLEWARE_KEY_PREFIX
+    cache_timeout = cache_timeout or settings.CACHE_MIDDLEWARE_SECONDS
+    cache = cache or caches[settings.CACHE_MIDDLEWARE_ALIAS]
+
+    cache_key = _generate_cache_header_key_s3_compatible(key_prefix, request)
+    headerlist = []
+    if response.has_header("Vary"):
+        is_accept_language_redundant = settings.USE_I18N
+        for header in cc_delim_re.split(response.headers["Vary"]):
+            header = header.upper().replace("-", "_")
+            if header != "ACCEPT_LANGUAGE" or not is_accept_language_redundant:
+                headerlist.append("HTTP_" + header)
+        headerlist.sort()
+
+    # if there is no Vary header, we still need a cache key
+    # for the request.build_absolute_uri()
+    cache.set(cache_key, headerlist, cache_timeout)
+    return _generate_cache_key_s3_compatible(
+        request, request.method, headerlist, key_prefix
+    )
+
+
+class CacheMiddlewareS3Compatible(CacheMiddleware):
+    """
+    Drop-in replacement for Django's CacheMiddleware that produces
+    cache keys compatible with S3ExpressCacheBackend.
+
+    This middleware automatically switches to S3-compatible key generation
+    when using `S3ExpressCacheBackend`.
+
+    For non-S3 backends, the middleware falls back to Django’s standard
+    cache key generation.
+    """
+
+    def _use_s3_backend(self) -> bool:
+        return isinstance(self.cache, S3ExpressCacheBackend)
```

Comment:
We're calling this on every request. I think the overhead of calling it is minimal, but it might be worth caching its value.

For example, in `__init__`:

self._is_s3_backend = isinstance(self.cache, S3ExpressCacheBackend)


Then, whenever we need it, we can just use the boolean value: `self._is_s3_backend`